### PR TITLE
Add validation for kube-scheduler configuration options

### LINKED ISSUE #66743: Add validation for kube-scheduler configuration options
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind technical-debt

[`KubeSchedulerConfiguration`](https://github.com/kubernetes/kubernetes/blob/28b6fb5f7d4ec9fc0be2843d61543ef70b748e70/pkg/apis/componentconfig/types.go#L79) does not have any validation at API level. We should add validation for the config options. We should also add validation for the [cli flags](https://github.com/kubernetes/kubernetes/blob/28b6fb5f7d4ec9fc0be2843d61543ef70b748e70/cmd/kube-scheduler/app/options/deprecated.go#L41).

/sig scheduling

### PR DESCRIPTION
**Pull request by noqcks**
**What this PR does / why we need it**: This adds validation to the kube-scheduler so that we're not accepting bogus values to the kube-scheduler. As requested by @bsalamat in issue https://github.com/kubernetes/kubernetes/issues/66743

**Which issue(s) this PR fixes**:
Fixes #66743

**Special notes for your reviewer**:
- Not sure if this validation is too heavy handed. Would love some feedback. 
- I started working on this before I realized @islinwb was also working on this same problem... https://github.com/kubernetes/kubernetes/pull/66787 I put this PR up anyways since I'm sure good code exists in both. I wasn't aware of the /assign command so didn't assign myself before starting work. 
- I didn't have time to work on adding validation to deprecated cli options. If the rest of this looks ok, I can finish that up.
- I hope the location of IsValidSocketAddr is correct. Lmk if it isn't. 

**Release note**:
```release-note
Adding validation to kube-scheduler at the API level
```

### CODE REVIEW

**Review by bsalamat — COMMENTED** [review]
Thanks, @noqcks!
Some early comments.

File: pkg/apis/componentconfig/validation/validation.go
Comment (neolit123) [inline pkg/apis/componentconfig/validation/validation.go]: `need to set atleast one of File or ConfigMap`
->
`needs to set either File or ConfigMap`
?

File: pkg/apis/componentconfig/validation/validation.go
Comment (neolit123) [inline pkg/apis/componentconfig/validation/validation.go]: `a positive`

File: pkg/apis/componentconfig/validation/validation.go
Comment (neolit123) [inline pkg/apis/componentconfig/validation/validation.go]: `a positive`

File: pkg/apis/componentconfig/validation/validation_test.go
Comment (neolit123) [inline pkg/apis/componentconfig/validation/validation_test.go]: 2018

**Comment by neolit123** [comment]
@noqcks thanks for the PR.
please make sure you run `./hack/update-bazel.sh` and include any file changes in this PR.

File: staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go
Comment (noqcks) [reply staging/src/k8s.io/apimachinery/pkg/util/validation/validation_test.go]: Added.

File: pkg/apis/componentconfig/validation/validation_test.go
Comment (noqcks) [reply pkg/apis/componentconfig/validation/validation_test.go]: I added some validations for deprecated options @bsalamat, but for some reason the function `ListAlgorithmProviders` was not returning the proper values I would expect "ClusterAutoscalerProvider | DefaultProvider" and was instead returning an empty string 🤔 . 

I'm not sure exactly how I could get this to work. I was planning on using `factory.ListAlgorithmProviders()` in the `Validate` function. If you have any idea of a way I could get the list of valid algorithm providers I would love to understand. Thanks.

**Comment by neolit123** [comment]
@noqcks 
hi, the bazel test is complaining about:

> I0731 23:35:06.761] /bazel-scratch/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/sandbox/linux-sandbox/4523/execroot/__main__/cmd/kube-scheduler/app/options/options.go:188:15: cannot use validation.ValidateKubeSchedulerConfiguration(&o.ComponentConfig) (type field.ErrorList) as type []error in append

**Comment by neolit123** [comment]
@noqcks 

the verify test indicates some small problems:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/66799/pull-kubernetes-verify/100514/:
```
pkg/apis/componentconfig/validation/validation.go:25:1: exported function ValidateKubeSchedulerConfiguration should have comment or be unexported
pkg/apis/componentconfig/validation/validation.go:48:1: exported function ValidateClientConnectionConfiguration should have comment or be unexported
pkg/apis/componentconfig/validation/validation.go:59:1: exported function 
....
```

**Comment by noqcks** [comment]
hmmm.... can't seem to figure out if this is an issue with my code or the test infrastructure. 

/retest

**Comment by neolit123** [comment]
@noqcks the e2e tests flake from time to time.

**Comment by noqcks** [comment]
@neolit123 not sure how I can debug these failed e2e tests 🤔I can't find anything in the logs that shows an error on my side. lmk if you have any ideas.

**Comment by neolit123** [comment]
@noqcks 
just babysit the tests with the `/test xxxxxx` command until they become green.

**Comment by noqcks** [comment]
I may try again a little later tonight. I checked a couple other PRs and they also seem to be failing e2e tests in the same manner. 

Thank you for help

File: pkg/apis/componentconfig/validation/validation.go
Comment (neolit123) [inline pkg/apis/componentconfig/validation/validation.go]: could be refactored as:

```
allErrs := field.ErrorList{}
if cc == nil {
	return append(allErrs, field.Invalid(fldPath.Child("policy"), nil, "need to set Policy"))
}
if cc.File == nil && cc.ConfigMap == nil {
	return append(allErrs, field.Invalid(fldPath.Child("file"), nil, "need to set either File or ConfigMap"))
}
return allErrs
```
to avoid indenting the `} else` part.

**Comment by noqcks** [comment]
@neolit123 `pull-kubernetes-kubemark-e2e-gce-big` seems to be running an older commit 🤔 Is there a way I can make it run on `31e73be`

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: We should validate leader election config only when leader election is enabled.

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: I am not sure if 0 is a valid Duration. If it is not, the error message at the next line should be changed.

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: similar to the previous comment.

File: pkg/apis/componentconfig/validation/validation.go
Comment (noqcks) [reply pkg/apis/componentconfig/validation/validation.go]: I think no as well: https://github.com/kubernetes/kubernetes/pull/64218/files

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: Actually, the error condition should remain as `cc.LeaseDuration.Duration <= 0` here and below.

**Comment by noqcks** [comment]
ready for another review on this I believe @bsalamat

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: the error message states that the value "must be non-negative". Non-negative includes zero, but zero causes an error. So, the error message should be changed to "must be positive" or "must be larger than zero". The same problem exists in the following checks as well.

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: When `cc.LeaderElect` is false, we should skip these two checks too.

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: `LeaseDuration` and `RenewDeadline` can be equal and that's not an error. Please replace <= by <.

File: pkg/apis/componentconfig/validation/validation_test.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation_test.go]: Could you please add test cases for renewDeadline and LeaseDuration being zero as well?

**Review by bsalamat — COMMENTED** [review]
Thanks, @noqcks! Mostly minor comments.

File: pkg/apis/componentconfig/validation/validation_test.go
Comment (noqcks) [reply pkg/apis/componentconfig/validation/validation_test.go]: added

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: oh, this condition is incorrect! This should actually be less than zero OR larger than 100:

```
if cc.HardPodAffinitySymmetricWeight < 0 || cc.HardPodAffinitySymmetricWeight > 100
```

Could you please add two cases to the unit tests, one for HardPodAffinitySymmetricWeight being less than zero and another one HardPodAffinitySymmetricWeight being larger than 100 and check that they fail validation?

File: pkg/apis/componentconfig/validation/validation.go
Comment (bsalamat) [inline pkg/apis/componentconfig/validation/validation.go]: nit: You should be able to do:
```
if !cc.LeaderElect {
```

**Review by bsalamat — COMMENTED** [review]
Thanks, @noqcks! Almost there!

File: pkg/apis/componentconfig/validation/validation.go
Comment (noqcks) [reply pkg/apis/componentconfig/validation/validation.go]: hmmm really? Isn't LeaderElect nested within the `LeaderElectionConfiguration` struct? Or is it magic in golang I don't know about

Comment (bsalamat) [reply pkg/apis/componentconfig/validation/validation.go]: That's how inheritance works in Go.

**Review by bsalamat — COMMENTED** [review]
/lgtm

File: cmd/kube-scheduler/app/options/options.go
Comment (smarterclayton) [inline cmd/kube-scheduler/app/options/options.go]: If I upgrade a cluster from 1.11, can a previously working scheduler start to fail because of validation?  Ie can this validation cause failures on upgrade?

Comment (noqcks) [reply cmd/kube-scheduler/app/options/options.go]: Sorry, im not familiar with the kubeadm upgrade process, but I would assume yes. 

@bsalamat should know more

Comment (bsalamat) [reply cmd/kube-scheduler/app/options/options.go]: I do not expect failures. The validations we are doing here are basic checks to ensure that the values are in the expected range. These validations would fail only when the provided config do not comply with the documented values in the API.

**Comment by noqcks** [comment]
@bsalamat had to make some changes because of an update to componentconfig/types.go in master

**Comment by noqcks** [comment]
still need review/approval from 

`pkg/apis/componentconfig/OWNERS`
/assign @sttts @mtaufen @vishh

& awaiting response from @smarterclayton

File: pkg/apis/componentconfig/validation/validation.go
Comment (sttts) [inline pkg/apis/componentconfig/validation/validation.go]: We shouldn't add apimachinery validations here. Put them where the types are.

**Comment by luxas** [comment]
/assign @luxas 
Thanks for this PR @noqcks, this is very important!
This PR needs to be rebased now that kube-scheduler types moved to staging, but I hope we can get this into v1.12. I'll review this ASAP.

**Review by bsalamat — APPROVED** [review]
/lgtm

**Comment by noqcks** [comment]
/test pull-kubernetes-integration

seems like this is flaking ?

**Comment by wilberusa** [comment]
unsubsribe

**Comment by noqcks** [comment]
@luxas ready for review

**Review by luxas — APPROVED** [review]
/lgtm
Thanks!

File: pkg/scheduler/apis/config/validation/validation.go
Comment (liggitt) [inline pkg/scheduler/apis/config/validation/validation.go]: where is this range documented or enforced today?

File: staging/src/k8s.io/apiserver/pkg/apis/config/validation/validation.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/apis/config/validation/validation.go]: is `ResourceLock` not required?

File: staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation.go]: this is not accurate... a negative number disables QPS rate limiting today

File: staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation_test.go
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation_test.go]: negative numbers are accepted today and disable client-side rate limiting. please add a test ensuring this continues to be accepted.

File: staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation.go
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation.go]: this is the only validation I see around the client config struct today:

https://github.com/kubernetes/kubernetes/blob/113872798d34daa32277616b99a368244f1f7b65/pkg/proxy/apis/config/validation/validation.go#L190-L194

that could be useful to promote to a reusable method

Comment (noqcks) [reply staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation.go]: Thanks, I'll fix. 

For Burst, is there any range that's not accepted?

File: pkg/scheduler/apis/config/validation/validation.go
Comment (noqcks) [reply pkg/scheduler/apis/config/validation/validation.go]: documented here: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/v1/types.go#L42

Comment (liggitt) [reply pkg/scheduler/apis/config/validation/validation.go]: please also add that to the flag help. if setting this to a value less than 0 or greater than 100 does not prevent the scheduler from working today, then I wouldn't make this a fatal error.

Comment (noqcks) [reply pkg/scheduler/apis/config/validation/validation.go]: added in 3f7830f6826b0ac8b5170f00014d4aaee3fb621f

File: staging/src/k8s.io/apiserver/pkg/apis/config/validation/validation.go
Comment (noqcks) [reply staging/src/k8s.io/apiserver/pkg/apis/config/validation/validation.go]: added in a5554d8dc7b46622ed8bb3c9fffe21a0ea247f94

File: staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation_test.go
Comment (noqcks) [reply staging/src/k8s.io/apimachinery/pkg/apis/config/validation/validation_test.go]: added in 125f64b

**Comment by noqcks** [comment]
@liggitt ready for another review

File: pkg/scheduler/apis/config/validation/validation.go
Comment (liggitt) [reply pkg/scheduler/apis/config/validation/validation.go]: >  if setting this to a value less than 0 or greater than 100 does not prevent the scheduler from working today, then I wouldn't make this a fatal error.

can you comment on this? @bsalamat, do you know what the effect of setting this to a value outside the 0-100 range is?

Comment (bsalamat) [reply pkg/scheduler/apis/config/validation/validation.go]: @liggitt Scheduler [has a check](https://github.com/kubernetes/kubernetes/blob/e1a270d6dcfd73e35db3970a042db9c704ec5a8f/pkg/scheduler/factory/factory.go#L1142) to ensure that this value is between 1 and 100. This is done at the scheduler creation. So, it was not possible to provide a value out of this range before this PR.

Comment (liggitt) [reply pkg/scheduler/apis/config/validation/validation.go]: Ok, great

Comment (liggitt) [reply pkg/scheduler/apis/config/validation/validation.go]: Does that mean the allowed range is 1-100 or 0-100?

**Comment by liggitt** [comment]
One last question on the 0-100 range, might need a doc update depending on what is actually allowed. Lgtm otherwise after a squash

File: pkg/scheduler/apis/config/validation/validation.go
Comment (bsalamat) [reply pkg/scheduler/apis/config/validation/validation.go]: 0 is changed to the default weight in our defaulting logic.

Comment (bsalamat) [reply pkg/scheduler/apis/config/validation/validation.go]: I am referring to this piece of code: https://github.com/kubernetes/kubernetes/blob/2d7b92ee743de20d17406003e463a829a0db5a51/pkg/scheduler/apis/config/v1alpha1/defaults.go#L43

**Comment by liggitt** [comment]
/approve

will leave lgtm to @bsalamat after squash

**Comment by noqcks** [comment]
squashed

**Review by bsalamat — COMMENTED** [review]
/lgtm

**Pull request merged**
